File: Src/ECScape.Wasm/Systems/InputSystem.cs
using ECScape.Components;
using ECScape.Engine;
using ECScape.Systems;

namespace ECScape.Wasm.Systems;

public sealed class InputSystem : ISystem
{
    private readonly HashSet<string> keysDown = [];

    public void RegisterKeyDown(string key)
    {
        keysDown.Add(key.ToLower());
    }

    public void RegisterKeyUp(string key)
    {
        keysDown.Remove(key.ToLower());
    }

    public void Update(World world, float deltaTime)
    {
        foreach (var entity in world.Entities
                     .Where(e => e.HasComponent<PlayerControllable>() && e.HasComponent<Velocity>()))
        {
            var velocity = entity.GetRequiredComponent<Velocity>();
            HandleInput(velocity);
        }
    }

    private void HandleInput(Velocity velocity)
    {
        if (keysDown.Contains("w") || keysDown.Contains("arrowup"))
        {
            velocity.Y = Configuration.PlayerMovementVertical;
        }
        if (keysDown.Contains("a") || keysDown.Contains("arrowleft"))
        {
            velocity.X = -Configuration.PlayerMovementHorizontal;
        }
        if (keysDown.Contains("d") || keysDown.Contains("arrowright"))
        {
            velocity.X = Configuration.PlayerMovementHorizontal;
        }
    }
}

File: Web/src/systems/InputSystem.js
import ISystem from './ISystem.js';
import { PlayerControllable, Velocity } from '../components/Components.js';
import Configuration from '../core/configuration.js';

class InputSystem extends ISystem {
    constructor() {
        super();
        this.keys = {};
        window.addEventListener('keydown', (e) => this.keys[e.key] = true);
        window.addEventListener('keyup', (e) => this.keys[e.key] = false);
    }

    update(world, deltaTime) {
        world.getEntitiesWith(PlayerControllable, Velocity).forEach(e => this.updateEntity(e));
    }

    updateEntity(entity) {
        this.handlePlayerInput(entity.getRequiredComponent(Velocity));
    }

    handlePlayerInput(velocity) {
        if (this.keys['w'] || this.keys['ArrowUp']) {
            velocity.Y = Configuration.PlayerMovementVertical;
        }
        if (this.keys['a'] || this.keys['ArrowLeft']) {
            velocity.X = -Configuration.PlayerMovementHorizontal;
        }
        if (this.keys['d'] || this.keys['ArrowRight']) {
            velocity.X = Configuration.PlayerMovementHorizontal;
        }
    }
}

export default InputSystem;